platform/linuxbsd/freedesktop_portal_desktop.cpp

```diff
@@ -617,6 +749,263 @@ void FreeDesktopPortalDesktop::_system_theme_changed_callback() {
+}
+
+bool FreeDesktopPortalDesktop::indicator_create(DisplayServer::IndicatorID p_id, const Ref<Image> &p_icon) {
+	MutexLock mutex_lock(file_dialog_mutex);
+
+	ERR_FAIL_COND_V(indicators.has(p_id), false);
+
+	if (unsupported) {
+		return false;
+	}
+
+	DBusError error;
+	dbus_error_init(&error);
+
+	DBusConnection *bus = dbus_bus_get(DBUS_BUS_SESSION, &error);
+	if (dbus_error_is_set(&error)) {
+		if (OS::get_singleton()->is_stdout_verbose()) {
+			ERR_PRINT(vformat("Error opening D-Bus connection: %s", error.message));
+		}
+		dbus_error_free(&error);
+		unsupported = true;
+		return false;
+	}
+
+	int pid = OS::get_singleton()->get_process_id();
+
+	String name = vformat("org.freedesktop.StatusNotifierItem-%d-%d", pid, p_id);
+	dbus_bus_request_name(bus, name.utf8(), DBUS_NAME_FLAG_DO_NOT_QUEUE, &error);
+	if (dbus_error_is_set(&error)) {
+		dbus_error_free(&error);
+		return false;
+	}
+
+	indicator_id_map[name] = p_id;
+
+	StatusNotifierItem &item = indicators[p_id];
+	item.id = p_id;
+	item.connection = bus;
+	item.name = name;
+
+	// TODO: Error handling.
+	indicator_set_icon(p_id, p_icon);
+
+	const char *name_ptr = name.utf8();
+
+	DBusMessage *message = dbus_message_new_method_call(
+			BUS_STATUS_NOTIFIER_WATCHER_NAME, BUS_STATUS_NOTIFIER_WATCHER_PATH, BUS_INTERFACE_STATUS_NOTIFIER_WATCHER,
+			"RegisterStatusNotifierItem");
+	dbus_message_append_args(
+			message,
+			DBUS_TYPE_STRING, &name_ptr,
+			DBUS_TYPE_INVALID);
+
+	dbus_connection_send(bus, message, NULL);
+	dbus_message_unref(message);
+
+	item.registered = true;
+
+	return true;
+}
+
+Error FreeDesktopPortalDesktop::indicator_set_icon(DisplayServer::IndicatorID p_id, const Ref<Image> &p_icon) {
+	MutexLock mutex_lock(file_dialog_mutex);
+
+	ERR_FAIL_COND_V(!indicators.has(p_id), ERR_UNCONFIGURED);
+	ERR_FAIL_COND_V(p_icon->get_size().x == 0, ERR_INVALID_PARAMETER);
+	ERR_FAIL_COND_V(p_icon->get_size().y == 0, ERR_INVALID_PARAMETER);
+
+	ERR_FAIL_COND_V(p_icon.is_null(), FAILED);
+
+	Ref<Image> image = p_icon->duplicate(true);
+
+	if (p_icon->is_compressed()) {
+		Error err = image->decompress();
+		ERR_FAIL_COND_V_MSG(err != OK, err, "Couldn't decompress VRAM-compressed status-icon. Switch to a lossless compression mode in the Import dock.");
+	}
+
+	// TODO: Figure out the size limits (at least swaybar errors on big pixmaps)
+	image->resize(64, 64);
+
+	StatusNotifierItem &item = indicators[p_id];
+
+	Vector<uint8_t> image_data = image->get_data();
+	item.icon_size = image->get_size();
+
+	// The API requires ARGB32 (or ARGB8888), which `Image` does not support...
+	// Sigh, we'll have to convert it badly ourselves.
+
+	item.icon_data.clear();
+
+	for (int y = 0; y < image->get_width(); ++y) {
+		for (int x = 0; x < image->get_width(); ++x) {
+			Color pixel = image->get_pixel(x, y);
+
+			item.icon_data.append(pixel.get_a8());
+			item.icon_data.append(pixel.get_r8());
+			item.icon_data.append(pixel.get_g8());
+			item.icon_data.append(pixel.get_b8());
+		}
+	}
```

Comment:
`get_pixel` have a bunch of unnecessary checks + color conversion to float `Color` and is pretty slow, so in the similar cases we usually use:

```
image->convert(Image::FORMAT_RGBA8);
const uint8_t *r = image->get_data().ptr();
```

and accessing RGBA8 data from `r` pointer directly.

Reply:
Names are most likely from GL - https://registry.khronos.org/OpenGL-Refpages/gl4/html/glTexImage2D.xhtml